### PR TITLE
Add a Semantic Header Phase to Painless Compiler

### PR DESCRIPTION
**Pull request by jdconrad**
This change creates a new phase that runs prior to the main semantic phase. This phase collects what would be considered header information such as user defined functions names. These are required prior to the main semantic phase; however, we were relying on the fact that we could easily cycle through the SFunction nodes in the user tree to collect these on our initial visit to the SClass user tree node. This is fragile and may not remain true in the future. This naturally separates this phase and allows for further information in the future to be analyzed such as user defined fields or even user defined objects.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/SemanticHeaderPhase.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Override
+    public Void visitFunction(SFunction userFunctionNode, ScriptScope scriptScope) {
+        String functionName = userFunctionNode.getFunctionName();
+        List<String> canonicalTypeNameParameters = userFunctionNode.getCanonicalTypeNameParameters();
+        List<String> parameterNames = userFunctionNode.getParameterNames();
+        int parameterCount = canonicalTypeNameParameters.size();
+
+        if (parameterCount != parameterNames.size()) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/SemanticHeaderPhase.java:49]: Check `canonicalTypeNameParameters.size()` != `parameterNames.size()`?

**Review by stu-elastic — APPROVED** [review]
Nice to have the as a pre-pass.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/SemanticHeaderPhase.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/SemanticHeaderPhase.java]: Good catch. Guessing this was left over from prior to the refactor where this was actually re-used more than once. Fixed.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review. Will commit once CI passes.

**Pull request merged**